### PR TITLE
[Ready for review] [1/X] Add volunteer signup [back-end]

### PR DESCRIPTION
**Pull request by nikhitashanker**
- Add new class for volunteer signup
- Alter database schema for addition of signup table
- Add database functions for signup 
   - Upon insert update the corresponding volunteering opportunity to reduce number of spots left by 1
   - Retrieve all signups by volunteering opportunity

### CODE REVIEW

**Comment by nikhitashanker** [comment]
Presubmit errors will be fixed by events PR.

File: project/src/main/java/com/google/sps/data/OpportunitySignup.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  private String opportunityId;
+  private String email;
+
+  public static class Builder {
+    // Required parameters
+    private String opportunityId;
+    private String email;
+
+    public Builder(String opportunityId, String email) {
+      this.opportunityId = opportunityId;
```
Comment (yicyao) [inline project/src/main/java/com/google/sps/data/OpportunitySignup.java:14]: Does opportunityId always = eventId? will this cause clashes if multiple opportunities per eventid?

File: project/src/test/java/com/google/sps/OpportunitySignupTasksTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    SpannerTestTasks.cleanup();
+  }
+
+  @Test
+  public void opportunitySignupInsertOpportunityUpdated() {
+    VolunteeringOpportunity opportunity =
+        new VolunteeringOpportunity.Builder(EVENT_ID, NAME, NUMBER_OF_SPOTS).build();
+    SpannerTasks.insertVolunteeringOpportunity(opportunity);
+
+    OpportunitySignup signup =
```
Comment (yicyao) [inline project/src/test/java/com/google/sps/OpportunitySignupTasksTest.java:44]: maybe a constant for `opportunity.getOpportunityId()`

File: project/src/main/java/com/google/sps/utilities/SpannerTasks.java [hunk 3]
```diff
@@ -292,4 +306,82 @@ public static void updateVolunteeringOpportunity(VolunteeringOpportunity opportu
...
+            new TransactionCallable<Void>() {
+              @Override
+              public Void run(TransactionContext transaction) throws Exception {
+                Struct row =
+                    transaction.readRow(
+                        VOLUNTEERING_OPPORTUNITY_TABLE,
+                        Key.of(signup.getOpportunityId()),
+                        Arrays.asList(NUM_SPOTS_LEFT));
+                long numSpotsLeft = row.getLong(0);
+                if (numSpotsLeft > 0) {
```
Comment (yicyao) [inline project/src/main/java/com/google/sps/utilities/SpannerTasks.java:342]: is it possible for numSpotsLeft == 0 or < 0? (Is it possible for someone to accidentally sign up to volunteer for an event w no opportunities left?)

File: project/src/test/java/com/google/sps/OpportunitySignupTasksTest.java (hunk 2 above)
Comment (nikhitashanker) [reply project/src/test/java/com/google/sps/OpportunitySignupTasksTest.java:44]: Done - thanks.

File: project/src/main/java/com/google/sps/utilities/SpannerTestTasks.java [hunk 4]
```diff
@@ -36,7 +36,11 @@ public static void setup() throws InterruptedException, ExecutionException {
                 + "  NumSpotsLeft               INT64 NOT NULL,"
                 + "  RequiredSkills             ARRAY<STRING(MAX)>,"
                 + "  Volunteers                 ARRAY<STRING(MAX)>"
-                + ") PRIMARY KEY (VolunteeringOpportunityID)");
+                + ") PRIMARY KEY (VolunteeringOpportunityID)",
```
Comment (dash102) [inline project/src/main/java/com/google/sps/utilities/SpannerTestTasks.java:39]: nit: Format with tabs so we can quickly see the columns and types

Comment (nikhitashanker) [reply project/src/main/java/com/google/sps/utilities/SpannerTestTasks.java:39]: Done.

**Pull request merged**